Don't call Dispatchers.Main getter during UnconfinedTestDispatcher construction

<!--
This is a place for issue reports that are not exactly bugs (wrong unintentional behavior) but for our library behaving suboptimally (though this could have been intentional).

**Double-check**

* If the behavior is strange, surprising, and undocumented, it could be a good idea to file a "Bug report" instead.
* Is this still relevant with the latest version of the library? We could have changed this already.
* Maybe there are good reasons for the existing behavior. Please try searching for existing discussions of the problem.
* Are you using the right abstraction? Consider asking on [StackOverflow](https://stackoverflow.com/) or the [Kotlin Slack](https://surveys.jetbrains.com/s3/kotlin-slack-sign-up). Maybe your need is better solved by some other abstraction entirely.

-->

I'm opening this issue as a result of my short conversation with @dkhalanskyjb on Kotlin Slack: https://kotlinlang.slack.com/archives/C1CFAFJSK/p1734007891082299

**What do we have now?**

In certain situations, creating an instance of `UnconfinedTestDispatcher` by calling `UnconfinedTestDispatcher()` just before setting it as the main dispatcher via `Dispatchers.setMain`, e.g. in a JUnit rule, may fail.

(The same applies to `StandardTestDispatcher`)

The execution path of creating the dispatcher was like this:

1. `UnconfinedTestDispatcher` https://github.com/Kotlin/kotlinx.coroutines/blob/6c6df2b850382887462eeaf51f21f58bd982491d/kotlinx-coroutines-test/common/src/TestCoroutineDispatchers.kt#L83
2. `TestMainDispatcher.currentTestScheduler` https://github.com/Kotlin/kotlinx.coroutines/blob/6c6df2b850382887462eeaf51f21f58bd982491d/kotlinx-coroutines-test/common/src/internal/TestMainDispatcher.kt#L50
3. `TestMainDispatcher.currentTestDispatcher` https://github.com/Kotlin/kotlinx.coroutines/blob/6c6df2b850382887462eeaf51f21f58bd982491d/kotlinx-coroutines-test/common/src/internal/TestMainDispatcher.kt#L47
4. `Dispatchers.Main` (getter invoked before `Dispatchers.setMain`!) https://github.com/Kotlin/kotlinx.coroutines/blob/6c6df2b850382887462eeaf51f21f58bd982491d/kotlinx-coroutines-core/jvm/src/Dispatchers.kt#L20
5. `MainDispatcherLoader.loadMainDispatcher`
6. `MainDispatchersKt.tryCreateDispatcher`
7. `kotlinx.coroutines.android.AndroidDispatcherFactory.createDispatcher`
8. `Looper.getMainLooper()` ⚡ throws 

<details><summary>Stacktrace</summary>

```
kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for DispatchedContinuation[Dispatchers.IO, Continuation at kotlinx.coroutines.flow.internal.ChannelFlow$collectToFun$1.invokeSuspend(ChannelFlow.kt:56)@6c0c966f]. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at app//kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:142)
	at app//kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:113)
	at app//kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:111)
	at app//kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:99)
	at app//kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
	at app//kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:811)
	at app//kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:715)
	at app//kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:702)
Caused by: java.lang.IllegalStateException: Module with the Main dispatcher had failed to initialize. For tests Dispatchers.setMain from kotlinx-coroutines-test module can be used
	at kotlinx.coroutines.internal.MissingMainCoroutineDispatcher.missing(MainDispatchers.kt:111)
	at kotlinx.coroutines.internal.MissingMainCoroutineDispatcher.isDispatchNeeded(MainDispatchers.kt:92)
	at kotlinx.coroutines.test.internal.TestMainDispatcher.isDispatchNeeded(TestMainDispatcher.kt:27)
	at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:156)
	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:466)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:500)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default(CancellableContinuationImpl.kt:489)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:364)
	at kotlinx.coroutines.channels.BufferedChannel$BufferedChannelIterator.tryResumeHasNextOnClosedChannel(BufferedChannel.kt:1737)
	at kotlinx.coroutines.channels.BufferedChannel.resumeWaiterOnClosedChannel(BufferedChannel.kt:2204)
	at kotlinx.coroutines.channels.BufferedChannel.resumeReceiverOnClosedChannel(BufferedChannel.kt:2191)
	at kotlinx.coroutines.channels.BufferedChannel.cancelSuspendedReceiveRequests(BufferedChannel.kt:2184)
	at kotlinx.coroutines.channels.BufferedChannel.completeClose(BufferedChannel.kt:1961)
	at kotlinx.coroutines.channels.BufferedChannel.isClosed(BufferedChannel.kt:2240)
	at kotlinx.coroutines.channels.BufferedChannel.isClosedForSend0(BufferedChannel.kt:2215)
	at kotlinx.coroutines.channels.BufferedChannel.isClosedForSend(BufferedChannel.kt:2212)
	at kotlinx.coroutines.channels.BufferedChannel.completeCloseOrCancel(BufferedChannel.kt:1933)
	at kotlinx.coroutines.channels.BufferedChannel.closeOrCancelImpl(BufferedChannel.kt:1826)
	at kotlinx.coroutines.channels.BufferedChannel.close(BufferedChannel.kt:1785)
	at kotlinx.coroutines.channels.SendChannel$DefaultImpls.close$default(Channel.kt:95)
	at kotlinx.coroutines.channels.ProducerCoroutine.onCompleted(Produce.kt:140)
	at kotlinx.coroutines.channels.ProducerCoroutine.onCompleted(Produce.kt:133)
	at kotlinx.coroutines.AbstractCoroutine.onCompletionInternal(AbstractCoroutine.kt:90)
	at kotlinx.coroutines.JobSupport.tryFinalizeSimpleState(JobSupport.kt:293)
	at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:867)
	at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:839)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:97)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	... 6 more
Caused by: java.lang.IllegalStateException: The main looper is not available
	at kotlinx.coroutines.android.AndroidDispatcherFactory.createDispatcher(HandlerDispatcher.kt:51)
	at kotlinx.coroutines.internal.MainDispatchersKt.tryCreateDispatcher(MainDispatchers.kt:53)
	at kotlinx.coroutines.test.internal.TestMainDispatcherFactory.createDispatcher(TestMainDispatcherJvm.kt:11)
	at kotlinx.coroutines.internal.MainDispatchersKt.tryCreateDispatcher(MainDispatchers.kt:53)
	at kotlinx.coroutines.internal.MainDispatcherLoader.loadMainDispatcher(MainDispatchers.kt:34)
	at kotlinx.coroutines.internal.MainDispatcherLoader.<clinit>(MainDispatchers.kt:18)
	at kotlinx.coroutines.Dispatchers.getMain(Dispatchers.kt:20)
	at kotlinx.coroutines.test.internal.TestMainDispatcher$Companion.getCurrentTestDispatcher$kotlinx_coroutines_test(TestMainDispatcher.kt:47)
	at kotlinx.coroutines.test.internal.TestMainDispatcher$Companion.getCurrentTestScheduler$kotlinx_coroutines_test(TestMainDispatcher.kt:50)
	at kotlinx.coroutines.test.TestCoroutineDispatchersKt.UnconfinedTestDispatcher(TestCoroutineDispatchers.kt:83)
	at kotlinx.coroutines.test.TestCoroutineDispatchersKt.UnconfinedTestDispatcher$default(TestCoroutineDispatchers.kt:79)
	at com.example.MainDispatcherRule.<init>(MainDispatcherRule.kt:14)
	at com.example.Test.<init>(Test.kt:51)
	(...)
```

</details> 

It prevented me from setting the `UnconfinedTestDispatcher` as `Main` because it wasn't possible to construct it in the first place. The error also wasn't helpful because it told me to call `setMain` which happened to be the exact thing I was trying to accomplish.

To be perfectly honest, I'm not sure how exactly to reproduce these circumstances. The project where I stumbled upon this problem was large and had a complicated setup. I wouldn't rule out the possibility of a user error or a misconfiguration of some sort, such as a version conflict in dependencies or a test classpath pollution (e.g. having `...-android` in the test classpath). Nevertheless, it would be great if this problem could be avoided somehow.

**What should be instead?**

If possible, the code necessary for substituting the main dispatcher (such as the `UnconfinedTestDispatcher()` or `StandardTestDispatcher()` function) should not call `Dispatchers.Main` (getter) because it leads to strange situations like the one I described where the user is told to call `setMain` even though that's exactly what the user is trying to accomplish.

**Why?**

_The upsides of your proposal._
* _Who would benefit from this and how?_ - People who accidentally ended up with coroutines-android in unit tests, I suppose
  - They wouldn't be told to do exactly the thing they are trying to do when they get the exception

**Why not?**

_The downsides of your proposal that you already see._
* _Is this a breaking change?_ - Probably not
* _Are there use cases that are better solved by what we have now?_ - I don't know
* _Does some code become less clear after this change?_ - Good question
* It was the only project where I found this problem and it makes me believe it's a user error after all.

--- a/kotlinx-coroutines-test/common/src/internal/TestMainDispatcher.kt
+++ b/kotlinx-coroutines-test/common/src/internal/TestMainDispatcher.kt
@@ -9,31 +9,36 @@ import kotlin.coroutines.*
  * The testable main dispatcher used by kotlinx-coroutines-test.
  * It is a [MainCoroutineDispatcher] that delegates all actions to a settable delegate.
  */
-internal class TestMainDispatcher(delegate: CoroutineDispatcher):
+internal class TestMainDispatcher(createInnerMain: () -> CoroutineDispatcher):
     MainCoroutineDispatcher(),
     Delay
 {
-    private val mainDispatcher = delegate
-    private var delegate = NonConcurrentlyModifiable(mainDispatcher, "Dispatchers.Main")
+    internal constructor(delegate: CoroutineDispatcher): this({ delegate })
+
+    private val mainDispatcher by lazy(createInnerMain)
+    private var delegate = NonConcurrentlyModifiable<CoroutineDispatcher?>(null, "Dispatchers.Main")
+
+    private val dispatcher
+        get() = delegate.value ?: mainDispatcher
 
     private val delay
-        get() = delegate.value as? Delay ?: defaultDelay
+        get() = dispatcher as? Delay ?: defaultDelay
 
     override val immediate: MainCoroutineDispatcher
-        get() = (delegate.value as? MainCoroutineDispatcher)?.immediate ?: this
+        get() = (dispatcher as? MainCoroutineDispatcher)?.immediate ?: this
 
-    override fun dispatch(context: CoroutineContext, block: Runnable) = delegate.value.dispatch(context, block)
+    override fun dispatch(context: CoroutineContext, block: Runnable) = dispatcher.dispatch(context, block)
 
-    override fun isDispatchNeeded(context: CoroutineContext): Boolean = delegate.value.isDispatchNeeded(context)
+    override fun isDispatchNeeded(context: CoroutineContext): Boolean = dispatcher.isDispatchNeeded(context)
 
-    override fun dispatchYield(context: CoroutineContext, block: Runnable) = delegate.value.dispatchYield(context, block)
+    override fun dispatchYield(context: CoroutineContext, block: Runnable) = dispatcher.dispatchYield(context, block)
 
     fun setDispatcher(dispatcher: CoroutineDispatcher) {
         delegate.value = dispatcher
     }
 
     fun resetDispatcher() {
-        delegate.value = mainDispatcher
+        delegate.value = null
     }
 
     override fun scheduleResumeAfterDelay(timeMillis: Long, continuation: CancellableContinuation<Unit>) =

--- a/kotlinx-coroutines-test/jvm/src/internal/TestMainDispatcherJvm.kt
+++ b/kotlinx-coroutines-test/jvm/src/internal/TestMainDispatcherJvm.kt
@@ -10,7 +10,21 @@ internal class TestMainDispatcherFactory : MainDispatcherFactory {
         val secondBestFactory = otherFactories.maxByOrNull { it.loadPriority } ?: MissingMainCoroutineDispatcherFactory
         /* Do not immediately create the alternative dispatcher, as with `SUPPORT_MISSING` set to `false`,
         it will throw an exception. Instead, create it lazily. */
-        return TestMainDispatcher({ secondBestFactory.tryCreateDispatcher(otherFactories) })
+        return TestMainDispatcher({
+            val dispatcher = try {
+                secondBestFactory.tryCreateDispatcher(otherFactories)
+            } catch (e: Throwable) {
+                reportMissingMainCoroutineDispatcher(e)
+            }
+            if (dispatcher.isMissing()) {
+                reportMissingMainCoroutineDispatcher(runCatching {
+                    // attempt to dispatch something to the missing dispatcher to trigger the exception.
+                    dispatcher.dispatch(dispatcher, Runnable { })
+                }.exceptionOrNull()) // can not be null, but it does not matter.
+            } else {
+                dispatcher
+            }
+        })
     }
 
     /**
@@ -25,4 +39,14 @@ internal actual fun Dispatchers.getTestMainDispatcher(): TestMainDispatcher {
     val mainDispatcher = Main
     require(mainDispatcher is TestMainDispatcher) { "TestMainDispatcher is not set as main dispatcher, have $mainDispatcher instead." }
     return mainDispatcher
+}
+
+private fun reportMissingMainCoroutineDispatcher(e: Throwable? = null): Nothing {
+    throw IllegalStateException(
+        "Dispatchers.Main was accessed when the platform dispatcher was absent " +
+            "and the test dispatcher was unset. Please make sure that Dispatchers.setMain() is called " +
+            "before accessing Dispatchers.Main and that Dispatchers.Main is not accessed after " +
+            "Dispatchers.resetMain().",
+        e
+    )
 }

--- a/kotlinx-coroutines-test/jvm/src/internal/TestMainDispatcherJvm.kt
+++ b/kotlinx-coroutines-test/jvm/src/internal/TestMainDispatcherJvm.kt
@@ -8,8 +8,9 @@ internal class TestMainDispatcherFactory : MainDispatcherFactory {
     override fun createDispatcher(allFactories: List<MainDispatcherFactory>): MainCoroutineDispatcher {
         val otherFactories = allFactories.filter { it !== this }
         val secondBestFactory = otherFactories.maxByOrNull { it.loadPriority } ?: MissingMainCoroutineDispatcherFactory
-        val dispatcher = secondBestFactory.tryCreateDispatcher(otherFactories)
-        return TestMainDispatcher(dispatcher)
+        /* Do not immediately create the alternative dispatcher, as with `SUPPORT_MISSING` set to `false`,
+        it will throw an exception. Instead, create it lazily. */
+        return TestMainDispatcher({ secondBestFactory.tryCreateDispatcher(otherFactories) })
     }
 
     /**

--- a/ui/kotlinx-coroutines-android/android-unit-tests/test/ordered/tests/FirstMockedMainTest.kt
+++ b/ui/kotlinx-coroutines-android/android-unit-tests/test/ordered/tests/FirstMockedMainTest.kt
@@ -35,7 +35,7 @@ open class FirstMockedMainTest : TestBase() {
             component.launchSomething()
             throw component.caughtException
         } catch (e: IllegalStateException) {
-            assertTrue(e.message!!.contains("Dispatchers.setMain from kotlinx-coroutines-test"))
+            assertTrue(e.message!!.contains("Dispatchers.setMain"))
         }
     }
 }

